# Assignment

<font face='georgia'>
    
   <h4><strong>What does tf-idf mean?</strong></h4>

   <p>    
Tf-idf stands for <em>term frequency-inverse document frequency</em>, and the tf-idf weight is a weight often used in information retrieval and text mining. This weight is a statistical measure used to evaluate how important a word is to a document in a collection or corpus. The importance increases proportionally to the number of times a word appears in the document but is offset by the frequency of the word in the corpus. Variations of the tf-idf weighting scheme are often used by search engines as a central tool in scoring and ranking a document's relevance given a user query.
</p>
    
   <p>
One of the simplest ranking functions is computed by summing the tf-idf for each query term; many more sophisticated ranking functions are variants of this simple model.
</p>
    
   <p>
Tf-idf can be successfully used for stop-words filtering in various subject fields including text summarization and classification.
</p>
    
</font>

<font face='georgia'>
    <h4><strong>How to Compute:</strong></h4>

Typically, the tf-idf weight is composed by two terms: the first computes the normalized Term Frequency (TF), aka. the number of times a word appears in a document, divided by the total number of words in that document; the second term is the Inverse Document Frequency (IDF), computed as the logarithm of the number of the documents in the corpus divided by the number of documents where the specific term appears.

 <ul>
    <li>
<strong>TF:</strong> Term Frequency, which measures how frequently a term occurs in a document. Since every document is different in length, it is possible that a term would appear much more times in long documents than shorter ones. Thus, the term frequency is often divided by the document length (aka. the total number of terms in the document) as a way of normalization: <br>

$TF(t) = \frac{\text{Number of times term t appears in a document}}{\text{Total number of terms in the document}}.$
</li>
<li>
<strong>IDF:</strong> Inverse Document Frequency, which measures how important a term is. While computing TF, all terms are considered equally important. However it is known that certain terms, such as "is", "of", and "that", may appear a lot of times but have little importance. Thus we need to weigh down the frequent terms while scale up the rare ones, by computing the following: <br>

$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}}.$
for numerical stabiltiy we will be changing this formula little bit
$IDF(t) = \log_{e}\frac{\text{Total  number of documents}} {\text{Number of documents with term t in it}+1}.$
</li>
</ul>

<br>
<h4><strong>Example</strong></h4>
<p>

Consider a document containing 100 words wherein the word cat appears 3 times. The term frequency (i.e., tf) for cat is then (3 / 100) = 0.03. Now, assume we have 10 million documents and the word cat appears in one thousand of these. Then, the inverse document frequency (i.e., idf) is calculated as log(10,000,000 / 1,000) = 4. Thus, the Tf-idf weight is the product of these quantities: 0.03 * 4 = 0.12.
</p>
</font>

## Task-1

<font face='georgia'>
    <h4><strong>1. Build a TFIDF Vectorizer & compare its results with Sklearn:</strong></h4>

<ul>
    <li> As a part of this task you will be implementing TFIDF vectorizer on a collection of text documents.</li>
    <br>
    <li> You should compare the results of your own implementation of TFIDF vectorizer with that of sklearns implemenation TFIDF vectorizer.</li>
    <br>
    <li> Sklearn does few more tweaks in the implementation of its version of TFIDF vectorizer, so to replicate the exact results you would need to add following things to your custom implementation of tfidf vectorizer:
       <ol>
        <li> Sklearn has its vocabulary generated from idf sroted in alphabetical order</li>
        <li> Sklearn formula of idf is different from the standard textbook formula. Here the constant <strong>"1"</strong> is added to the numerator and denominator of the idf as if an extra document was seen containing every term in the collection exactly once, which prevents zero divisions.
            
 $IDF(t) = 1+\log_{e}\frac{1\text{ }+\text{ Total  number of documents in collection}} {1+\text{Number of documents with term t in it}}.$
        </li>
        <li> Sklearn applies L2-normalization on its output matrix.</li>
        <li> The final output of sklearn tfidf vectorizer is a sparse matrix.</li>
    </ol>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer.</li>
        <li> Print out the alphabetically sorted voacb after you fit your data and check if its the same as that of the feature names from sklearn tfidf vectorizer. </li>
        <li> Print out the idf values from your implementation and check if its the same as that of sklearns tfidf vectorizer idf values. </li>
        <li> Once you get your voacb and idf values to be same as that of sklearns implementation of tfidf vectorizer, proceed to the below steps. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> After completing the above steps, print the output of your custom implementation and compare it with sklearns implementation of tfidf vectorizer.</li>
        <li> To check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it.</li>
        </ol>
    </li>
    <br>
   </ul>

  <p> <font color="#e60000"><strong>Note-1: </strong></font> All the necessary outputs of sklearns tfidf vectorizer have been provided as reference in this notebook, you can compare your outputs as mentioned in the above steps, with these outputs.<br>
   <font color="#e60000"><strong>Note-2: </strong></font> The output of your custom implementation and that of sklearns implementation would match only with the collection of document strings provided to you as reference in this notebook. It would not match for strings that contain capital letters or punctuations, etc, because sklearn version of tfidf vectorizer deals with such strings in a different way. To know further details about how sklearn tfidf vectorizer works with such string, you can always refer to its official documentation.<br>
   <font color="#e60000"><strong>Note-3: </strong></font> During this task, it would be helpful for you to debug the code you write with print statements wherever necessary. But when you are finally submitting the assignment, make sure your code is readable and try not to print things which are not part of this task.
    </p>

### Corpus

In [1]:
## SkLearn# Collection of string documents

corpus = [
     'this is the first document',
     'this document is the second document',
     'and this is the third one',
     'is this the first document',
]

### SkLearn Implementation

In [2]:
from sklearn.feature_extraction.text import TfidfVectorizer
vectorizer = TfidfVectorizer()
vectorizer.fit(corpus)
skl_output = vectorizer.transform(corpus)

In [3]:
# sklearn feature names, they are sorted in alphabetic order by default.

print(vectorizer.get_feature_names())

['and', 'document', 'first', 'is', 'one', 'second', 'the', 'third', 'this']


In [4]:
# Here we will print the sklearn tfidf vectorizer idf values after applying the fit method
# After using the fit function on the corpus the vocab has 9 words in it, and each has its idf value.

print(vectorizer.idf_)

[1.91629073 1.22314355 1.51082562 1.         1.91629073 1.91629073
 1.         1.91629073 1.        ]


In [5]:
# shape of sklearn tfidf vectorizer output after applying transform method.

skl_output.shape

(4, 9)

In [136]:
# sklearn tfidf values for first line of the above corpus.
# Here the output is a sparse matrix

print(skl_output[0])

  (0, 8)	0.38408524091481483
  (0, 6)	0.38408524091481483
  (0, 3)	0.38408524091481483
  (0, 2)	0.5802858236844359
  (0, 1)	0.46979138557992045


In [7]:
# sklearn tfidf values for first line of the above corpus.
# To understand the output better, here we are converting the sparse output matrix to dense matrix and printing it.
# Notice that this output is normalized using L2 normalization. sklearn does this by default.

print(skl_output[0].toarray())

[[0.         0.46979139 0.58028582 0.38408524 0.         0.
  0.38408524 0.         0.38408524]]


### Your custom implementation

In [156]:
# Write your code here.
# Make sure its well documented and readble with appropriate comments.
# Compare your results with the above sklearn tfidf vectorizer
# You are not supposed to use any other library apart from the ones given below

from collections import Counter
from tqdm import tqdm
from scipy.sparse import csr_matrix
import math
import operator
from sklearn.preprocessing import normalize
import numpy

<p style="background-color:powderblue"> My Code <br>
- Split all the sentences to words <br>
- count all words <br>
- Find TF <br>
- Find IDF <br>

In [157]:
corpus                 # glancing corpus

['slow moving aimless movie distressed drifting young man',
 'not sure lost flat characters audience nearly half walked',
 'attempting artiness black white clever camera angles movie disappointed became even ridiculous acting poor plot lines almost non existent',
 'little music anything speak',
 'best scene movie gerardo trying find song keeps running head',
 'rest movie lacks art charm meaning emptiness works guess empty',
 'wasted two hours',
 'saw movie today thought good effort good messages kids',
 'bit predictable',
 'loved casting jimmy buffet science teacher',
 'baby owls adorable',
 'movie showed lot florida best made look appealing',
 'songs best muppets hilarious',
 'cool',
 'right case movie delivers everything almost right face',
 'average acting main person low budget clearly see',
 'review long overdue since consider tale two sisters single greatest film ever made',
 'put gem movie terms screenplay cinematography acting post production editing directing aspect film makin

In [158]:
type(corpus)        # is it a list

list

In [159]:
len(corpus)        # number of sentences

746

<p style="background-color:pink">
We have 746 sentences. Let's create a function which divides each sentence into individual words.

### Finding all unique words

In [160]:
BoW = []
for i in range(len(corpus)):
    BoW +=  corpus[i].split(" ")

In [161]:
print(BoW)

['slow', 'moving', 'aimless', 'movie', 'distressed', 'drifting', 'young', 'man', 'not', 'sure', 'lost', 'flat', 'characters', 'audience', 'nearly', 'half', 'walked', 'attempting', 'artiness', 'black', 'white', 'clever', 'camera', 'angles', 'movie', 'disappointed', 'became', 'even', 'ridiculous', 'acting', 'poor', 'plot', 'lines', 'almost', 'non', 'existent', 'little', 'music', 'anything', 'speak', 'best', 'scene', 'movie', 'gerardo', 'trying', 'find', 'song', 'keeps', 'running', 'head', 'rest', 'movie', 'lacks', 'art', 'charm', 'meaning', 'emptiness', 'works', 'guess', 'empty', 'wasted', 'two', 'hours', 'saw', 'movie', 'today', 'thought', 'good', 'effort', 'good', 'messages', 'kids', 'bit', 'predictable', 'loved', 'casting', 'jimmy', 'buffet', 'science', 'teacher', 'baby', 'owls', 'adorable', 'movie', 'showed', 'lot', 'florida', 'best', 'made', 'look', 'appealing', 'songs', 'best', 'muppets', 'hilarious', 'cool', 'right', 'case', 'movie', 'delivers', 'everything', 'almost', 'right', 'f

In [162]:
BoW_sorted = sorted(list(set(BoW)))
print(BoW_sorted)

['aailiyah', 'abandoned', 'ability', 'abroad', 'absolutely', 'abstruse', 'abysmal', 'academy', 'accents', 'accessible', 'acclaimed', 'accolades', 'accurate', 'accurately', 'accused', 'achievement', 'achille', 'ackerman', 'act', 'acted', 'acting', 'action', 'actions', 'actor', 'actors', 'actress', 'actresses', 'actually', 'adams', 'adaptation', 'add', 'added', 'addition', 'admins', 'admiration', 'admitted', 'adorable', 'adrift', 'adventure', 'advise', 'aerial', 'aesthetically', 'affected', 'affleck', 'afraid', 'africa', 'afternoon', 'age', 'aged', 'ages', 'ago', 'agree', 'agreed', 'aimless', 'air', 'aired', 'akasha', 'akin', 'alert', 'alexander', 'alike', 'allison', 'allow', 'allowing', 'almost', 'along', 'alongside', 'already', 'also', 'although', 'always', 'amateurish', 'amaze', 'amazed', 'amazing', 'amazingly', 'america', 'american', 'americans', 'among', 'amount', 'amusing', 'amust', 'anatomist', 'angel', 'angela', 'angeles', 'angelina', 'angle', 'angles', 'angry', 'anguish', 'angus

### Calculating TF
Sklearn has its vocabulary generated from idf sroted in alphabetical order

In [163]:
sorted(BoW_sorted)

['aailiyah',
 'abandoned',
 'ability',
 'abroad',
 'absolutely',
 'abstruse',
 'abysmal',
 'academy',
 'accents',
 'accessible',
 'acclaimed',
 'accolades',
 'accurate',
 'accurately',
 'accused',
 'achievement',
 'achille',
 'ackerman',
 'act',
 'acted',
 'acting',
 'action',
 'actions',
 'actor',
 'actors',
 'actress',
 'actresses',
 'actually',
 'adams',
 'adaptation',
 'add',
 'added',
 'addition',
 'admins',
 'admiration',
 'admitted',
 'adorable',
 'adrift',
 'adventure',
 'advise',
 'aerial',
 'aesthetically',
 'affected',
 'affleck',
 'afraid',
 'africa',
 'afternoon',
 'age',
 'aged',
 'ages',
 'ago',
 'agree',
 'agreed',
 'aimless',
 'air',
 'aired',
 'akasha',
 'akin',
 'alert',
 'alexander',
 'alike',
 'allison',
 'allow',
 'allowing',
 'almost',
 'along',
 'alongside',
 'already',
 'also',
 'although',
 'always',
 'amateurish',
 'amaze',
 'amazed',
 'amazing',
 'amazingly',
 'america',
 'american',
 'americans',
 'among',
 'amount',
 'amusing',
 'amust',
 'anatomist',
 'an

In [164]:
range(len(set(lst[0])))

range(0, 5)

In [165]:
TF_wholedoc = []
def TF(x):
    freq ={}
    for j in range(len(x)):         # looping over a unique wtokens in a sentence
        k = 0
        for i in range(len(x)):     # taking each token and checking over its entire sentence               
                if x[j] == x[i]:          
                    k += 1          # counting number of times a token is repeated
        d = {x[j]:(k/len(x))}       # assigning token and its frequency to a dictionary
        freq.update(d)
    return(TF_wholedoc.append(freq))
    del freq
for i in range(len(corpus)):
    TF(corpus[i].split(" "))

In [166]:
print(TF_wholedoc)
print(type(TF_wholedoc))

[{'slow': 0.125, 'moving': 0.125, 'aimless': 0.125, 'movie': 0.125, 'distressed': 0.125, 'drifting': 0.125, 'young': 0.125, 'man': 0.125}, {'not': 0.1111111111111111, 'sure': 0.1111111111111111, 'lost': 0.1111111111111111, 'flat': 0.1111111111111111, 'characters': 0.1111111111111111, 'audience': 0.1111111111111111, 'nearly': 0.1111111111111111, 'half': 0.1111111111111111, 'walked': 0.1111111111111111}, {'attempting': 0.05263157894736842, 'artiness': 0.05263157894736842, 'black': 0.05263157894736842, 'white': 0.05263157894736842, 'clever': 0.05263157894736842, 'camera': 0.05263157894736842, 'angles': 0.05263157894736842, 'movie': 0.05263157894736842, 'disappointed': 0.05263157894736842, 'became': 0.05263157894736842, 'even': 0.05263157894736842, 'ridiculous': 0.05263157894736842, 'acting': 0.05263157894736842, 'poor': 0.05263157894736842, 'plot': 0.05263157894736842, 'lines': 0.05263157894736842, 'almost': 0.05263157894736842, 'non': 0.05263157894736842, 'existent': 0.05263157894736842}


<class 'list'>


### Implementing IDF

In [167]:
import math
BoW_sorted

['aailiyah',
 'abandoned',
 'ability',
 'abroad',
 'absolutely',
 'abstruse',
 'abysmal',
 'academy',
 'accents',
 'accessible',
 'acclaimed',
 'accolades',
 'accurate',
 'accurately',
 'accused',
 'achievement',
 'achille',
 'ackerman',
 'act',
 'acted',
 'acting',
 'action',
 'actions',
 'actor',
 'actors',
 'actress',
 'actresses',
 'actually',
 'adams',
 'adaptation',
 'add',
 'added',
 'addition',
 'admins',
 'admiration',
 'admitted',
 'adorable',
 'adrift',
 'adventure',
 'advise',
 'aerial',
 'aesthetically',
 'affected',
 'affleck',
 'afraid',
 'africa',
 'afternoon',
 'age',
 'aged',
 'ages',
 'ago',
 'agree',
 'agreed',
 'aimless',
 'air',
 'aired',
 'akasha',
 'akin',
 'alert',
 'alexander',
 'alike',
 'allison',
 'allow',
 'allowing',
 'almost',
 'along',
 'alongside',
 'already',
 'also',
 'although',
 'always',
 'amateurish',
 'amaze',
 'amazed',
 'amazing',
 'amazingly',
 'america',
 'american',
 'americans',
 'among',
 'amount',
 'amusing',
 'amust',
 'anatomist',
 'an

In [168]:
IDF = []
for i in range(len(BoW_sorted)):
    # print(i)
    n = 0
    den = 0
    for j in range(len(corpus)):
        if BoW_sorted[i] in corpus[j].split(" "):
            n += 1
    idf = 1 + math.log(((1+len(corpus))/(1+n)))
    IDF.append(idf)

### Implementing TF-IDF vectorizer

In [181]:
dictionary = dict(zip(BoW_sorted,IDF))  # create a dictionary of IDF as values and tokens as keys
print(len(dictionary))

2897


In [174]:
TF_IDF = []
for i in range(len(corpus)):    # for every sentence (4)
    z = 0
    LIST = []
    for k in dictionary:        # print(k,dictionary[k])
        
        if k in TF_wholedoc[i]:
            LIST.append(TF_wholedoc[i][k]*dictionary[k])
        else:
            LIST.append(0)
        z += 1
        # print(LIST)
    TF_IDF.append(LIST)

In [175]:
normalized_TF_IDF = normalize(TF_IDF, norm='l2')
print(normalized_TF_IDF)

[[0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]]


In [177]:
CSR_Matrix = csr_matrix(normalized_TF_IDF,(len(corpus),len(BoW)))                # Converting the matrix to sparse matrix
print(CSR_Matrix.shape)
print(CSR_Matrix[0])

(746, 7610)
  (0, 53)	0.4123943870778812
  (0, 689)	0.4123943870778812
  (0, 721)	0.4123943870778812
  (0, 1549)	0.30566026894803877
  (0, 1655)	0.16192317905848022
  (0, 1657)	0.35781145622317734
  (0, 2294)	0.3377679916467555
  (0, 2888)	0.35781145622317734


In [ ]:
top_50 = CSR_Matrix.arg

## But the order is different :(
<p style="background-color:powderblue">Still a good progress. Understood clearly TF-IDF calculations for a corpus

## Task-2

<font face='georgia'>
    <h4><strong>2. Implement max features functionality:</strong></h4>

<ul>
    <li> As a part of this task you have to modify your fit and transform functions so that your vocab will contain only 50 terms with top idf scores.</li>
    <br>
    <li>This task is similar to your previous task, just that here your vocabulary is limited to only top 50 features names based on their idf values. Basically your output will have exactly 50 columns and the number of rows will depend on the number of documents you have in your corpus.</li>
    <br>
    <li>Here you will be give a pickle file, with file name <strong>cleaned_strings</strong>. You would have to load the corpus from this file and use it as input to your tfidf vectorizer.</li>
    <br>
    <li>Steps to approach this task:
    <ol>
        <li> You would have to write both fit and transform methods for your custom implementation of tfidf vectorizer, just like in the previous task. Additionally, here you have to limit the number of features generated to 50 as described above.</li>
        <li> Now sort your vocab based in descending order of idf values and print out the words in the sorted voacb after you fit your data. Here you should be getting only 50 terms in your vocab. And make sure to print idf values for each term in your vocab. </li>
        <li> Make sure the output of your implementation is a sparse matrix. Before generating the final output, you need to normalize your sparse matrix using L2 normalization. You can refer to this link https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.normalize.html </li>
        <li> Now check the output of a single document in your collection of documents,  you can convert the sparse matrix related only to that document into dense matrix and print it. And this dense matrix should contain 1 row and 50 columns. </li>
        </ol>
    </li>
    <br>
   </ul>

In [155]:
# Below is the code to load the cleaned_strings pickle file provided
# Here corpus is of list type

import pickle
with open('cleaned_strings', 'rb') as f:
    corpus = pickle.load(f)
    
# printing the length of the corpus loaded
print("Number of documents in corpus = ",len(corpus))

Number of documents in corpus =  746


In [ ]:
# Write your code here.
# Try not to hardcode any values.
# Make sure its well documented and readble with appropriate comments.

In [184]:
len(BoW)

7610

In [187]:
dictionary2 = {}
for i in BoW:
    x = BoW.count(i)
   # print(x)
    dictionary2.update({i:x})
print(dictionary2)

{'slow': 6, 'moving': 4, 'aimless': 1, 'movie': 182, 'distressed': 1, 'drifting': 1, 'young': 4, 'man': 13, 'not': 177, 'sure': 3, 'lost': 4, 'flat': 2, 'characters': 35, 'audience': 5, 'nearly': 1, 'half': 6, 'walked': 2, 'attempting': 1, 'artiness': 1, 'black': 10, 'white': 8, 'clever': 5, 'camera': 10, 'angles': 2, 'disappointed': 6, 'became': 2, 'even': 40, 'ridiculous': 5, 'acting': 43, 'poor': 6, 'plot': 28, 'lines': 7, 'almost': 9, 'non': 5, 'existent': 1, 'little': 17, 'music': 14, 'anything': 8, 'speak': 3, 'best': 25, 'scene': 14, 'gerardo': 1, 'trying': 4, 'find': 10, 'song': 4, 'keeps': 2, 'running': 2, 'head': 3, 'rest': 4, 'lacks': 6, 'art': 13, 'charm': 2, 'meaning': 3, 'emptiness': 1, 'works': 5, 'guess': 5, 'empty': 2, 'wasted': 7, 'two': 13, 'hours': 4, 'saw': 12, 'today': 4, 'thought': 12, 'good': 58, 'effort': 1, 'messages': 1, 'kids': 8, 'bit': 5, 'predictable': 10, 'loved': 10, 'casting': 6, 'jimmy': 2, 'buffet': 1, 'science': 1, 'teacher': 1, 'baby': 1, 'owls': 1

In [192]:
top50 = sorted(dictionary2, key=dictionary2.get, reverse=True)[:50]

In [193]:
print(top50)

['movie', 'not', 'film', 'one', 'bad', 'good', 'like', 'acting', 'time', 'really', 'great', 'even', 'see', 'would', 'characters', 'well', 'no', 'ever', 'movies', 'plot', 'story', 'made', 'could', 'best', 'also', 'films', 'character', 'script', 'seen', 'way', 'love', 'make', 'watching', 'think', 'real', 'every', 'much', 'work', 'look', 'funny', 'scenes', 'actors', 'better', 'cast', 'never', 'wonderful', 'little', 'watch', 'show', 'everything']
